vehicle/mkvi/software/imu/imu.c

```diff
@@ -0,0 +1,168 @@
+#include "vehicle/mkvi/software/imu/icm20948/icm20948.h"
+
+#include <avr/interrupt.h>
+#include <avr/io.h>
+#include <stdint.h>
+#include <util/delay.h>
+
+// #include "projects/btldr/btldr_lib.h"
+// #include "projects/btldr/git_sha.h"
+// #include "projects/btldr/libs/image/api.h"
+// #include "projects/can_print/can_print.h"
+
+// /*
+//  * Required for btldr
+//  */
+// image_hdr_t image_hdr __attribute__((section(".image_hdr"))) = {
+//     .image_magic = IMAGE_MAGIC,
+//     .git_sha = STABLE_GIT_COMMIT,
+// };
+
+/**
+ * Timer 0 for sending IMU data over CAN
+ */
+volatile bool can_send_imu_data = true;
+void timer_0_isr(void) {
+    can_send_imu_data = true;
+}
+
+/**
+ * Timer 1 for the LED heartbeat
+ */
+volatile bool led_heartbeat = true;
+void timer_1_isr(void) {
+    led_heartbeat = true;
+}
+
+uint8_t test = 0;
+/**
+ * Update the arb id of an IMU based on soldered pull up jumpers.
+ *
+ * Allows for multiple IMUs running the same firmware on the same CAN bus.
+ */
+void update_can_arb_id(void) {
+    // Pull up inputs
+    gpio_set_mode(bin_id_0, INPUT);
+    gpio_set_mode(bin_id_1, INPUT);
+    gpio_set_mode(bin_id_2, INPUT);
+
+    // Update ids
+    imu_accel_msg.id
+        += (0x10
+            * (!gpio_get_pin(bin_id_0) + (!gpio_get_pin(bin_id_1) << 1)
+               + (!gpio_get_pin(bin_id_2) << 2)));
+    imu_gyro_msg.id
+        += (0x10
+            * (!gpio_get_pin(bin_id_0) + (!gpio_get_pin(bin_id_1) << 1)
+               + (!gpio_get_pin(bin_id_2) << 2)));
+    imu_magnet_msg.id
+        += (0x10
+            * (!gpio_get_pin(bin_id_0) + (!gpio_get_pin(bin_id_1) << 1)
+               + (!gpio_get_pin(bin_id_2) << 2)));
+}
+
+/**
+ * Initialize 16m1 hardware peripherals
+ */
+void init_peripherals(void) {
+    gpio_set_mode(debug_led, OUTPUT); // Heartbeat LED
+    update_can_arb_id(); // Update arb based on IMU id
+    can_init_imu(); // CAN
+    sei(); // Interrupts
+    timer_init(&timer_0_cfg); // Timer 0
+    timer_init(&timer_1_cfg); // Timer 1
+    spi_init(&imu_spi_cfg); // SPI
+}
+
+/**
+ * Initialize ICM20948 for full 9-axis readings
+ */
+void init_imu(void) {
+    // Perform soft reset and wait for power up
+    icm_write_register(PWR_MGMT_1, 0x81);
+    uint8_t pwr_mgmt_1_data = 0;
+    while (pwr_mgmt_1_data != 0x41) {
+        icm_read_register(PWR_MGMT_1, &pwr_mgmt_1_data);
+    }
+
+    // Wake from sleep
+    icm_write_register(PWR_MGMT_1, 0x01);
+
+    switch_register_bank(BANK_2);
+
+    // Set accelerometer full scale range
+    icm_write_register(ACCEL_CONFIG, 0x1 | (accel_fsr << 1));
+
+    // Set gyroscope full scale range
+    icm_write_register(GYRO_CONFIG_1, 0x01 | (gyro_fsr << 1));
+
+    switch_register_bank(BANK_0);
+
+    // Initialize magnetometer
+    // init_magnetometer();
+}
+
+/**
+ * Read accel data and place in CAN structs
+ */
+void read_accel_data(void) {
+    uint8_t accel_data[2] = { 0x0, 0x0 };
+    icm_read_register(ACCEL_XOUT_L, &accel_data[0]);
```

Comment:
from my very very quick scan, I think the chip supports multibyte reads?
<img width="674" alt="Screenshot 2024-10-05 at 10 52 48 AM" src="https://github.com/user-attachments/assets/ed0d0a0c-8c2f-4b64-bf99-bc4111b698d1">

Reply:
Yes, and you can also use the built in FIFO buffer in the IMU to optimize even further. The SPI interface wasn't a bottleneck so I didn't get to it. Could be more useful in the future with a more powerful MCU.